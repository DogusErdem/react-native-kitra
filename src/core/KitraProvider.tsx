import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
