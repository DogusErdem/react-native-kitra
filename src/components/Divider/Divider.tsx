import { StyleSheet, View, ViewProps } from 'react-native';
import type { DividerProps, FCCWD } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
  },
});
const Divider : FCCWD<DividerProps&ViewProps> = (
  { theme,
    borderStyle = 'solid',
    width = '100%',
    color = theme?.primary30,
    style,
    ...props },
) => (
  <View style={[styles.container, { width }, style]} {...props}>
    <View style={{ width, borderWidth: StyleSheet.hairlineWidth, borderColor: color, borderStyle }} />
  </View>
);

export default applyDefaults(Divider);
