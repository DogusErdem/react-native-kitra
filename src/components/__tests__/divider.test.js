import { render } from '@testing-library/react-native';
import { KitraProvider } from '../../core/KitraProvider';
import Divider from '../Divider/Divider';

/* eslint-disable no-undef */

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Divider', () => {
  it('default render', () => {
    const tree = renderWithContext(<Divider />);
    expect(tree).toMatchSnapshot();
  });
  describe('render with custom theme', () => {
    it('render with default theme', () => {
      const { getByTestId } = renderWithContext(<Divider theme={{ default: { background: 'blue' } }} />);
      const divider = getByTestId('divider');
      expect(divider).toHaveStyle({ borderColor: 'blue' });
    });
  });
  describe('render with different style and prop', () => {
    it('render with width prop', () => {
      const { getByTestId } = renderWithContext(<Divider width={120} />);
      const container = getByTestId('dividerContainer');
      const divider = getByTestId('divider');
      expect(container).toHaveStyle({ width: 120 });
      expect(divider).toHaveStyle({ width: 120 });
    });
    it('render with borderWidth prop', () => {
      const { getByTestId } = renderWithContext(<Divider borderWidth={2} width={110} />);
      const container = getByTestId('dividerContainer');
      const divider = getByTestId('divider');
      expect(container).toHaveStyle({ width: 110 });
      expect(divider).toHaveStyle({ width: 110, borderWidth: 2 });
    });
    it('render with borderWidth prop', () => {
      const { getByTestId } = renderWithContext(<Divider
        borderWidth={2}
        width={110}
        style={{ marginTop: 10, padding: 15 }}
      />);
      const container = getByTestId('dividerContainer');
      const divider = getByTestId('divider');
      expect(container).toHaveStyle({ width: 110, marginTop: 10, padding: 15 });
      expect(divider).toHaveStyle({ width: 110, borderWidth: 2 });
    });
  });
  describe('render with different variant', () => {
    it('render with solid variant', () => {
      const { getByTestId } = renderWithContext(<Divider variant="solid" />);
      const divider = getByTestId('divider');
      expect(divider).toHaveStyle({ borderStyle: 'solid' });
    });
    it('render with solid dashed', () => {
      const { getByTestId } = renderWithContext(<Divider variant="dashed" />);
      const divider = getByTestId('divider');
      expect(divider).toHaveStyle({ borderStyle: 'dashed' });
    });
    it('render with solid dotted', () => {
      const { getByTestId } = renderWithContext(<Divider variant="dotted" />);
      const divider = getByTestId('divider');
      expect(divider).toHaveStyle({ borderStyle: 'dotted' });
    });
  });
});
