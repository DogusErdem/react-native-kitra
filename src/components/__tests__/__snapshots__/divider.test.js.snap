// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider default render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "justifyContent": "center",
        },
        {
          "width": "100%",
        },
        undefined,
      ]
    }
    testID="dividerContainer"
    typography={
      {
        "body": {
          "medium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 14,
            "lineHeight": 21,
          },
          "regular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 14,
            "lineHeight": 16,
          },
          "smedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 12,
            "lineHeight": 18,
          },
          "sregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 12,
            "lineHeight": 16,
          },
          "xsmedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 10,
            "lineHeight": 16,
          },
          "xsregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 10,
            "lineHeight": 14,
          },
          "xxsmedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 8,
            "lineHeight": 12,
          },
          "xxsregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 8,
            "lineHeight": 12,
          },
        },
        "heading": {
          "h1": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 30,
            "lineHeight": 45,
          },
          "h2": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 28,
            "lineHeight": 42,
          },
          "h3": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 23,
            "lineHeight": 35,
          },
          "h4": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 19,
            "lineHeight": 29,
          },
          "h5": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 16,
            "lineHeight": 25,
          },
        },
      }
    }
  >
    <View
      style={
        [
          {
            "justifyContent": "center",
          },
          {
            "borderColor": "#195CEF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "width": "100%",
          },
        ]
      }
      testID="divider"
    />
  </View>
</View>
`;
